Resolve "Web FTP"

geunwoonoh opened the pull request:
As mentioned in https://github.com/alpacanetworks/alpamon-go/issues/1, the following changes have been made to natively replace the FTP function with AIP:

- Implement FTPClient
   - An FTPClient has been implemented to establish a Control Connection with alpacon-server.
   - FTPClient receives Web FTP Commands, which map to FTP Commands, from the user and processes them.

FTP Command | Web FTP Command
-- | --
LIST | list
MKD | mkd
CWD | cwd
PWD | pwd
DELE | dele
RMD | rmd
RNFR/RNTO | mv
Not Defined | cp

- `openftp` Command Addition
   - `openftp` Command runs `FTPClient` as go routine.
   - `openftp` command integrate into the `CommandRunner`.

- Update upload & download alpamon Command
   - Web FTP provides FTP's Data Connection-like functionality through a REST API.
   - The upload and download commands support the following features.
      - bulk upload/download
      - recursive upload/download
      - zip upload
      - unzip download

Closes #1

geunwoonoh committed:
Add openftp Command

Add openftp Command which is responsible
for launching FtpClient as goroutine.

geunwoonoh committed:
Add FtpClient

Add FtpClient to handle Web FTP commands which
corresponding FTP Command.
It creates a pty with permissions matching the
username and groupname, and executes logic
 based on the specific Web FTP command.

geunwoonoh committed:
Add setFtpCmdSysProcAttrAndEnv

Add setFtpCmdSysProcAttrAndEnv() function to
configure SysProcAttr and env for cmd.
Differentiate between macOS and Linux.

geunwoonoh committed:
Update convertGroupIds() to ConvertGroupIds()

convertGroupIds() function from pty_linux.go is now
used in setFtpCmdSysProcAttrAndEnv() at
ftp_linux.go.
To make it globally accessible, it has been moved to
utils.go and renamed to ConvertGroupIds().
As a result, pty_linux.go has been modified accordingly.

geunwoonoh committed:
Update runFileDownload()

To provide bulk and recursive functionality during
file uploads, runFileDownload() function has been
updated.
To achieve this, parsePaths() and makeArchive()
functions have been added.

geunwoonoh committed:
Add github.com/google/uuid package

Add github.com/google/uuid package

geunwoonoh committed:
Minor fix

Change cleanPath to parsePath

geunwoonoh committed:
Update runFileDownload()

To provide bulk and recursive functionality during
file uploads, runFileDownload() function has been
updated.
To achieve this, fileDownload() and isZipFile()
functions have been added.

geunwoonoh committed:
Add File struct

Add Files key to CommandData structure
and File structure to support bulk uploads.

geunwoonoh committed:
Refactor bulk upload in makeArchive()

Refactor the bulk upload process to prevent
the root directory from being included in the
compressed file.

geunwoonoh committed:
Refactor Web FTP Commands

Web FTP Commands has been refactored
to operate within a subprocess with limited privileges.
This change was implemented to enhance security for
users executing Web FTP commands, and it was achieved
by using the os/exec package.
Unused fields and functions were also removed.

geunwoonoh committed:
Delete unused files

Unused files were deleted due to changes in
Web FTP command's functionality.

geunwoonoh committed:
Updated error code return mechanism

Update error code return logic to accommodate
the new approach of directly executing Linux
commands using the os/exec package.

geunwoonoh committed:
Add check permission logic in cwd Command

Add check permission logic in cwd Command.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 1-web-ftp

royroyee commented on pkg/runner/ftp.go:
```diff
@@ -0,0 +1,428 @@
+var ftpTerminals map[string]*FtpClient
+
+func init() {
+	ftpTerminals = make(map[string]*FtpClient)
```
A global variable `terminals` was added in pty for `session_id` reference due to resize, but if FTP doesn’t require this, please remove the global variable

royroyee committed:
Minor fix

geunwoonoh committed:
Minor fix

Apply Golang linting.

royroyee commented on pkg/runner/command.go:
```diff
@@ -468,7 +489,22 @@ func (cr *CommandRunner) runFileUpload(fileName string) (exitCode int, result st
+		return 1, err.Error()
+	}
+	if bulk || recursive {
+		defer os.Remove(name)
```
Ensure errors are handled explicitly when using defer (Go convention). Review and fix other areas as needed
```go	
	defer func() { _ = os.Remove(name) }()
```

geunwoonoh committed:
Merge remote-tracking branch 'origin/1-web-ftp' into 1-web-ftp

royroyee commented on pkg/runner/command.go:
```diff
@@ -586,3 +632,108 @@ func getFileData(data CommandData) ([]byte, error) {
 	return content, nil
 }
+
+func parsePaths(pathList []string) ([]string, bool, bool, error) {
```
Consider adding descriptive names for return values to improve readability, especially when there are multiple return variables in Go. For example:
```go
func (cr *CommandRunner) delGroup() (exitCode int, result string) {

```

royroyee commented on pkg/runner/ftp_types.go:
```diff
@@ -0,0 +1,89 @@
+	Error   map[string]int `json:"error"`
+}
+
+var returnCodes = map[string]returnCode{
```
It’s better to use const for the error messages rather than repeating them:
```go
const (
    ErrPermissionDenied = "permission denied"
    ErrInvalidArgument  = "invalid argument"
    ErrNoSuchFile       = "no such file or directory"
    ...
    )
```
Additionally, instead of returning `map[string]string`, it would be clearer to define a specific type for the return value.

royroyee replied:
Permissioned -> Permission

geunwoonoh replied:
I've fixed typo!

geunwoonoh committed:
Fix space to ('// ')

To enable the shell to interpret spaces,
substituted double slashes ('//') for spaces.

geunwoonoh committed:
Add Web FTP related types

Add explicit structs to improve type safety
in FtpClient.
Define custom structures tailored to specific
data types, minimizing the frequency of type assertions.

geunwoonoh committed:
Update due to the addition of structs

Update data types by adding explicit structs
to ftp_types.go.

geunwoonoh committed:
Add install_zip_package to postinstall.sh

Add install_zip_package function to check
for the existence of zip package during
Alpamon installation and install it accordingly.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 1-web-ftp

royroyee commented on pkg/runner/ftp_types.go:
```diff
@@ -0,0 +1,136 @@
+	Command FtpCommand  `json:"command"`
+	Success bool        `json:"success"`
+	Code    int         `json:"code,omitempty"`
+	Data    interface{} `json:"data,omitempty"`
```
Consider defining a specific type for the `data` field instead of using an interface. Please review

royroyee committed:
Minor fix

royroyee commented on scripts/postinstall.sh:
```diff
@@ -24,6 +25,32 @@ check_systemd_status() {
+  fi
+
+  if command -v apt-get &> /dev/null; then
+    apt-get update
```
No update needed. it’s already included in the Linux package script

geunwoonoh committed:
Apply code review

Apply go lang convention.
Fix makeArchive() to declare cmd once.
Add exception handling for when there are
no elements in cr.data.Paths.

geunwoonoh committed:
Apply code review

Delete FtpTerminals.
Change FtpResult.Data's type to CommandResult.
Add Const for error message.
Update ftp.go due to modification FtpResult.

geunwoonoh committed:
Minor fix

Delete apt-get update

geunwoonoh committed:
Fix makeArchive()

Resolving issues caused by refactoring

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 1-web-ftp

geunwoonoh committed:
Merge remote-tracking branch 'origin/1-web-ftp' into 1-web-ftp

geunwoonoh committed:
Minor fix

Apply go lang convention

geunwoonoh committed:
Refactor makeArchive()

Previously, use cd command to navigate to
the working directory, but when executing in
sh -c environment, replace spaces in the path
with "// " for compatibility. This reduced readability.
To address this, implemente a solution using cmd.Dir
and cmd.Args to change the working directory,
allowing us to input paths with spaces directly.

geunwoonoh committed:
Minor fix

Fix typo

royroyee commented on pkg/runner/ftp.go:
```diff
@@ -0,0 +1,412 @@
+func (fc *FtpClient) list(rootDir string, depth int) (CommandResult, error) {
+	path := fc.parsePath(rootDir)
+	cmdResult, err := fc.listRecursive(path, depth, 0)
+	return *cmdResult, err
```
Remove unnecessary pointer dereferencing

royroyee committed:
Fix minor code formatting

geunwoonoh committed:
Fix unnecessary pointer dereferencing

Fix unnecessary pointer dereferencing in list()
and listRecursive() functions.
